ext/random/engine_mt19937.c

```diff
@@ -0,0 +1,344 @@
+	/* Legacy mode deliberately not inside php_mt_rand_range()
+	 * to prevent other functions being affected */
+	RAND_RANGE_BADSCALING(r, min, max, PHP_MT_RAND_MAX);
+	return (zend_long) r;
+}
+
+static bool serialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_mt19937 *s = status->state;
+	zval t;
+	uint32_t i;
+
+	for (i = 0; i< MT_N; i++) {
+		ZVAL_STR(&t, php_random_bin2hex_le(&s->state[i], sizeof(uint32_t)));
+		zend_hash_next_index_insert(data, &t);
+	}
+	ZVAL_LONG(&t, s->count);
+	zend_hash_next_index_insert(data, &t);
+	ZVAL_LONG(&t, s->mode);
+	zend_hash_next_index_insert(data, &t);
+	
+	return true;
+}
+
+static bool unserialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_mt19937 *s = status->state;
+	zval *t;
+	uint32_t i;
+
+	for (i = 0; i < MT_N; i++) {
+		t = zend_hash_index_find(data, i);
+		if (!t || Z_TYPE_P(t) != IS_STRING || Z_STRLEN_P(t) != (2 * sizeof(uint32_t))) {
+			return false;
+		}
+		if (!php_random_hex2bin_le(Z_STR_P(t), &s->state[i])) {
+			return false;
+		}
+	}
+	t = zend_hash_index_find(data, MT_N); 
+	if (!t || Z_TYPE_P(t) != IS_LONG) {
+		return false;
+	}
+	s->count = Z_LVAL_P(t);
+	t = zend_hash_index_find(data, MT_N + 1);
+	if (!t || Z_TYPE_P(t) != IS_LONG) {
+		return false;
+	}
+	s->mode = Z_LVAL_P(t);
+
+	return true;
+}
+
+const php_random_algo php_random_algo_mt19937 = {
+	sizeof(uint32_t),
+	sizeof(php_random_status_state_mt19937),
+	seed,
+	generate,
+	range,
+	serialize,
+	unserialize
+};
+
+/* {{{ php_random_mt19937_seed_default */
+PHPAPI void php_random_mt19937_seed_default(php_random_status_state_mt19937 *state)
+{
+	zend_long seed = 0;
+
+	if (php_random_bytes_silent(&seed, sizeof(zend_long)) == FAILURE) {
+		seed = GENERATE_SEED();
+	}
+
+	mt19937_seed_state(state, (uint64_t) seed);
+}
+/* }}} */
+
+/* {{{ Random\Engine\Mt19937::__construct() */
+PHP_METHOD(Random_Engine_Mt19937, __construct)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	php_random_status_state_mt19937 *state = engine->status->state;
+	zend_long seed, mode = MT_RAND_MT19937;
+	bool seed_is_null = true;
+
+	ZEND_PARSE_PARAMETERS_START(0, 2)
+		Z_PARAM_OPTIONAL;
+		Z_PARAM_LONG_OR_NULL(seed, seed_is_null);
+		Z_PARAM_LONG(mode);
+	ZEND_PARSE_PARAMETERS_END();
+
+	switch (mode) {
+		case MT_RAND_PHP:
+			state->mode = MT_RAND_PHP;
+			break;
+		default:
+			state->mode = MT_RAND_MT19937;
+	}
+
+	if (seed_is_null) {
+		/* MT19937 has a very large state, uses CSPRNG for seeding only */
+		if (php_random_bytes_silent(&seed, sizeof(zend_long)) == FAILURE) {
+			zend_throw_exception(spl_ce_RuntimeException, "Random number generation failed", 0);
+			RETURN_THROWS();
+		}
+	}
+
+	engine->algo->seed(engine->status, seed);
+}
+/* }}} */
+
+/* {{{ Random\Engine\Mt19937::generate() */
+PHP_METHOD(Random_Engine_Mt19937, generate)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	uint64_t generated;
+	size_t size;
+	zend_string *bytes;
+	int i;
```

Comment:
we no longer require declaring vars at the beginning of function as C99 is used so you can just do it below when you assign value...

Reply:
fixed: 3ba7af86e414663085ce902f4f6a594535c9d631
Thanks!